CMakeLists.txt

```diff
@@ -0,0 +1,127 @@
+cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+
+project(CppExtraSamples)
+
+set(CMAKE_CXX_STANDARD 11)
+set(ZIVID_VERSION 1.5.0)
+
+option(USE_EIGEN3 "Enable samples which depend on Eigen 3" ON)
+option(USE_OPENCV "Enable samples which depend on OpenCV" ON)
+option(USE_PCL "Enable samples which depend on Point Cloud Library (PCL)" ON)
+option(USE_VIS3D "Enable samples which depend on the Zivid Cloud Visualizer. These can be hard to get to build on Linux due to a specific Qt dependency" ON)
+
+find_package(Zivid ${ZIVID_VERSION} COMPONENTS Core REQUIRED)
+find_package(PCL 1.2 QUIET)
+find_package(OpenCV 4.0.1 QUIET OPTIONAL_COMPONENTS core highgui)
+
+set(SAMPLES
+    Zivid/CaptureHDRCompleteSettings
+    Zivid/CaptureHDRLoop
+    Zivid/CaptureSavePLY
+    Zivid/ConnectToSerialNumberCamera
+    Zivid/GetCameraIntrinsics
+    Zivid/ReadZDF
+    Zivid/ZDF2PLY
+    ZividPCL/ReadPCLVis3D
+    ZividPCL/CaptureWritePCLVis3D
+    ZividPCL/CaptureFromFileWritePCLVis3D
+    ZividPCL/ZDF2PCD
+    ZividOpenCV/ZDF2OpenCV
+    Downsample)
+
+set(Eigen3_DEPENDING Downsample)
+set(PCL_DEPENDING ReadPCLVis3D CaptureWritePCLVis3D CaptureFromFileWritePCLVis3D ZDF2PCD)
+set(OpenCV_DEPENDING ZDF2OpenCV)
+set(Vis3D_DEPENDING Downsample CaptureFromFileWritePCLVis3D CaptureWritePCLVis3D ZDF2OpenCV CaptureHDRLoop CaptureHDRCompleteSettings ReadZDF)
+
+macro(disable_samples DEPENDENCY_NAME)
+    message("${DEPENDENCY_NAME} samples have been disabled:")
+    foreach(SAMPLE ${SAMPLES})
+        get_filename_component(SAMPLE_NAME ${SAMPLE} NAME)
+        set(DEPENDENCY_LIST "${DEPENDENCY_NAME}_DEPENDING")
+        if(${SAMPLE_NAME} IN_LIST ${DEPENDENCY_LIST})
+            message("  - ${SAMPLE}")
+            list(REMOVE_ITEM SAMPLES ${SAMPLE})
+        endif()
+    endforeach()
+endmacro()
+
+if(USE_EIGEN3)
+    if(NOT DEFINED EIGEN3_INCLUDE_DIR)
```

Comment:
Why don't you wait with find_package(.. REQUIRED ..) until here? Then you would get the official error message to the consumer and you dont have to invent your own, that needs to be kept in sync with Eigen's config/find script.

Reply:
Eigen is not using `find_package`,  I assume you're referring to PCL and OpenCV? That's a good point.

Reply:
I agree that it's a good idea to move `find_package` inside `if(USE....)`. And I assume you're referring to `PCL` and `OpenCV` only, since `Eigen` is not found by a CMake script.

In that case: There's nothing in those error messages which needs to be kept in sync with config/find scripts. `OpenCV_DIR` and `PCL_DIR` are not inputs to the OpenCV and PCL config files, they are arguments to CMake to know where to *look for* those config files in the first place. So I don't think this is actually a problem.

There's also another very valuable thing in the custom error messages: Letting the user know they can turn the dependency off with `-DUSE_THETHING=OFF`. So I think we should keep those messages.

My suggestion then is to move `find_package` inside `if(USE_...)`, but don't add `REQUIRED`. You can then both see CMake's verbose message about not finding a config file, *and* our own message.

What do you think @nedrebo and @SatjaSivcev ?

Reply:
Made a draft PR #26 so you can see what I'm proposing.

Reply:
#26 looks good

Reply:
Cool, I added a better description and un-drafted it now. Thanks for the feedback! :)